### ISSUE TITLE
ConfigurationAccessors for DependenciesHandlerScope and PluginContainer

### ISSUE THREAD

**Issue by driver733** [gradle/gradle#12737]
It would be great if we could use ConfigurationAccessors, such as `ImplementationConfigurationAccessors`, `KaptConfigurationAccessors`, etc., in Gradle Kotlin DSL. This would allow to use the syntax that is used in the `build.gradle` in Gradle plugin development.

### Expected Behavior
```
 override fun apply(project: Project) {
        with(project) {
            with(plugins) {
                apply<KotlinPlatformJvmPlugin>()
           }
           dependencies {
                implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
                api("com.github.pozo:mapstruct-kotlin:1.3.1.1")
                kapt("org.mapstruct:mapstruct-jdk8:1.3.1.Final")
                testImplementation("org.jetbrains.kotlin:kotlin-test")
           }
}
```

### Current Behavior
```
 override fun apply(project: Project) {
        with(project) {
            with(plugins) {
                apply(KotlinPlatformJvmPlugin::class::java)
           }
           dependencies {
                "implementation"("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
                "api"("com.github.pozo:mapstruct-kotlin:1.3.1.1")
                "kapt"("org.mapstruct:mapstruct-jdk8:1.3.1.Final")
                "testImplementation"("org.jetbrains.kotlin:kotlin-test")
           }
}
```

### Context
Implementation of a Gradle Plugin in Kotlin using Gradle Kotlin DSL.

### [Example](https://github.com/driver733/gradle-kotlin-setup-plugin/blob/master/src/main/kotlin/com/driver733/gradle/plugin/GradleKotlinSetupPlugin.kt#L18)

**Comment by vlsi** [comment]
Have you tried [precompiled script plugins](https://docs.gradle.org/current/userguide/kotlin_dsl.html#kotdsl:precompiled_plugins)?

See samples in JUnit5
* declaration https://github.com/junit-team/junit5/blob/7868b9710c62892b7f4eaa93db6cec46722dfcd5/buildSrc/src/main/kotlin/publishing-conventions.gradle.kts
* usage: https://github.com/junit-team/junit5/blob/7868b9710c62892b7f4eaa93db6cec46722dfcd5/junit-bom/junit-bom.gradle.kts#L3

**Comment by vlsi** [comment]
At the plugin build time, Gradle does not know the project would have `implementation`, `kapt` and similar configurations.

How do you think it would know which accessors to generate?

On the other hand, if you use `precompiled script plugin`, then you could put the following in your `plugin` code:

```kotlin
plugins {
    `java-library`
..
}
```

Gradle would generate accessors appropriately.

So I'm inclined this is should be closed :-/

**Comment by driver733** [comment]
@vlsi I think the issue lies somewhere else. Gradle does generate the accessors. I can use them in the `{PROJECT_DIR}/build.gradle.kts`:

```
dependencies {
    implementation(kotlin("stdlib-jdk8"))
    implementation(kotlin("gradle-plugin"))
}
```

However, I cannot use the `DependencyHandler` extensions, such as  `ImplementationConfigurationAcceesors` shown above in the `{PROJECT_DIR}/src/main/kotlin/MyCustomGradlePlugin.kt`:

```
class GradleKotlinSetupPlugin : Plugin<Project> {

    override fun apply(project: Project) {
        with(project) {
           dependencies {
              "implementation"("org.jetbrains.kotlin:kotlin-stdlib-jdk8") // notice the quote
          }
}
```

All of the guides on the internet that describe how to implement a custom Gradle Plugin in Kotlin use the same notation.

For example, this [guide](https://proandroiddev.com/sharing-build-logic-with-kotlin-dsl-203274f73013) uses it.


My question is the following: Why can't I use the extension functions that are available in `{PROJECT_DIR}/build.gradle.kts` in `{PROJECT_DIR}/src/main/kotlin/MyCustomGradlePlugin.kt`?

**Comment by vlsi** [comment]
> My question is the following: Why can't I use the extension functions that are available in {PROJECT_DIR}/build.gradle.kts in {PROJECT_DIR}/src/main/kotlin/MyCustomGradlePlugin.kt?

Please read the very first sentence in https://github.com/gradle/gradle/issues/12737#issuecomment-611406339
It does answer exactly the question you ask.

TL;DR: rename `MyCustomGradlePlugin.kt` to `my-custom-plugin.gradle.kts`
It would enable accessors and it would make the plugin code easier to read.

**Comment by driver733** [comment]
@vlsi Thanks, I think I got it.

One more question: Is it possible to publish the plugin that is defined as a `precompiled script plugin`? The `PluginDeclaration` requires the `implementationClass` field to be set. What should I set it to? I am puzzled since the  `precompiled script plugin` does not define any classes...

**Comment by vlsi** [comment]
>Is it possible to publish the plugin that is defined as a precompiled script plugin? 

Of course, it is.

I think you could cut this block as `kotlin-dsl` plugin generates plugin descriptor automatically. It does generate regular plugin class, and it generates the needed property file that mentiones the class behind the scenes.

**Comment by driver733** [comment]
@vlsi Thanks. I have finally managed to make it work.

It would be great if this feature was documented on [this](https://docs.gradle.org/current/userguide/custom_plugins.html) page.

**Comment by big-guy** [comment]
Please open a new issue if you think we're missing something. We now have samples here that use precompiled script plugins in both Groovy and Kotlin: https://docs.gradle.org/current/samples/index.html
